consensus/src/dag/dag_network.rs

```diff
@@ -28,5 +39,136 @@ pub trait DAGNetworkSender: Send + Sync {
         responders: Vec<Author>,
         message: ConsensusMsg,
         timeout: Duration,
-    ) -> anyhow::Result<ConsensusMsg>;
+    ) -> RpcWithFallback;
+}
+
+struct Responders {
+    peers: Vec<Author>,
+    generator: ExponentialNumberGenerator,
+}
+
+impl Responders {
+    fn new(mut peers: Vec<Author>, initial_request_count: u32, max_request_count: u32) -> Self {
+        peers.shuffle(&mut rand::thread_rng());
+        Self {
+            peers,
+            generator: ExponentialNumberGenerator::new(initial_request_count, 2, max_request_count),
+        }
+    }
+
+    fn next_to_request(&mut self) -> Option<Vec<Author>> {
+        let count = self.generator.next().expect("should return a number");
+
+        if self.peers.is_empty() {
+            return None;
+        }
+        Some(
+            self.peers.split_off(
+                self.peers
+                    .len()
+                    .checked_sub(count as usize)
+                    .unwrap_or(0),
+            ),
+        )
+    }
+}
+
+pub struct RpcWithFallback {
+    responders: Responders,
+    message: ConsensusMsg,
+    timeout: Duration,
+
+    terminated: bool,
```

Comment:
You are not using this.
Where is the logic for the fetcher to kill the RPC when data is received?